### PR TITLE
Make `FunctionsAPI` use  `IAM.SessionsAPI` (remove separate implementation)

### PR DESCRIPTION
**Pull request by haakonvt**
### As posted on slack:
Hey there! I’m unable to call a function (it fails with `status="Failed"` and log message: `Could not fetch a valid token or a valid API key. Unable to instantiate a CogniteClient. Please verify your credentials`), depending on what type of `CredentialProvider` (from `cognite.client.credentials` ) I have used to instantiate my `CogniteClient`.

- OAuthClientCredentials ✅ 
- OAuthInteractive ❌ 
- OAuthDeviceCode ❌ 
- Token ✅ 
- User-made subclasses of `CredentialProvider` ❌ 

(Note: _All of these credentials are using the exact same app registration with all the required acls_)
Upon inspecting the code, it seems you only do an `insinstance`-check (on the ✅  ones only) and call the `SessionsAPI` _directly_ instead of using `client.iam.sessions[.create]` . I have ~~started a draft~~ [PR](https://github.com/cognitedata/cognite-sdk-python/pull/1136) to remove this “additional API implementation” and reduce coupling in the code base; a user is allowed to create his own `CredentialProvider` and that should work with Functions ootb imo 😄 

### Changes

- This PR makes both `OAuthInteractive`, `OAuthDeviceCode` (and any custom subclass impl.) "turn green" ✅ 
- The error message no longer includes whether "token exchange" or "client credentials" were used, but I think that's okay since we get the _actual_ error from the API instead. Probably that has some more concrete feedback to the user about what went wrong.
- Passing client credentials as a `ClientCredentials` object is now supported (extending the current behavior of passing a `dict` with two required keys)

## ~~TODO:~~
~~Tests are failing~~

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

**Review by ngerstle-cognite — COMMENTED** [review]
Does this remove support for using token exchange based sessions with functions in the python sdk?

**Comment by haakonvt** [comment]
> Does this remove support for using token exchange based sessions with functions in the python sdk?

@ngerstle-cognite quite the opposite in fact. With this change, all `CredentialProvider`s work, instead of just the 2 specific classes they checked for in the old code 😄

File: cognite/client/_api/functions.py
Comment (haakonvt) [inline cognite/client/_api/functions.py:501]: Note: `handle` is the wrong parameter name, thus changed to `function_handle`

File: CHANGELOG.md
Comment (haakonvt) [inline CHANGELOG.md:22]:
```suggestion
- `FunctionsAPI.call` now also works for clients using auth flow `OAuthInteractive`, `OAuthDeviceCode`, and any user-made subclass of `CredentialProvider`.
```

File: tests/tests_unit/test_api/test_functions.py
Comment (haakonvt) [inline tests/tests_unit/test_api/test_functions.py:570]:
```suggestion
        self, mock_functions_call_responses, cognite_client_with_client_credentials_flow
```

File: cognite/client/_api_client.py
Comment (erlendvollset) [inline cognite/client/_api_client.py:82]: Not make it Optional instead? Do we ever not pass it?

File: tests/tests_integration/test_api/test_transformations/test_transformations.py
Comment (erlendvollset) [inline tests/tests_integration/test_api/test_transformations/test_transformations.py:410]:
```suggestion
    @pytest.mark.xfail  # TODO: view and edgeType are mutually exclusive (code: 400)
```

File: cognite/client/_api_client.py
Comment (haakonvt) [reply cognite/client/_api_client.py:82]: Agreed, seems like a much better solution. If this causes a lot of `mypy`-cascading issues, I'll put it in a separate PR.

File: tests/tests_integration/test_api/test_transformations/test_transformations.py
Comment (haakonvt) [reply tests/tests_integration/test_api/test_transformations/test_transformations.py:410]: I guess this was already fixed in a separate PR. I'll verify.

File: cognite/client/_api_client.py
Comment (haakonvt) [reply cognite/client/_api_client.py:82]: ...yup, definitely separate PR 😂 

image

Comment (haakonvt) [reply cognite/client/_api_client.py:82]: Addressed in this PR: https://github.com/cognitedata/cognite-sdk-python/pull/1175

File: tests/tests_integration/test_api/test_transformations/test_transformations.py
Comment (haakonvt) [reply tests/tests_integration/test_api/test_transformations/test_transformations.py:410]: Also this now: https://github.com/cognitedata/cognite-sdk-python/pull/1174

**Comment by haakonvt** [comment]
@erlendvollset now that @geircognite has added certificate auth to the SDK, this PR suddenly doesn't add support for "all credential" types anymore (given that `SessionsAPI` does not support it). Should I add a check for it in `_create_session_and_return_nonce` (a function only used inside the `FunctionsAPI`) or just let it fail with the error message from the API (what I have done)?

Maybe you can say something about whether this might be supported soon/later/ever @ngerstle-cognite ? 😄

**Comment by geircognite** [comment]
> 

@haakonvt My preference would be to just add a check and then return a sensible error message if it is certificate auth. Just adding this as a convenience mainly for extractors and I don't see why those would need to use the sessions API.

**Review by qTipTip — APPROVED** [review]
This looks great!

Thanks a lot for this! Very nice to get rid of the arguably slightly hacky code to determine what type of login-flow was being used.

**Comment by haakonvt** [comment]
@erlendvollset I think I need an additional approval from you or someone "global" codeowners

**Review by erlendvollset — APPROVED** [review]

**Pull request merged**
